examples/static_set/host_bulk_perfect_example.cu

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuco/static_set.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/functional.h>
+#include <thrust/logical.h>
+#include <thrust/sequence.h>
+
+#include <iostream>
+#include <limits>
+
+/**
+ * @file host_bulk_example.cu
+ * @brief Demonstrates usage of the static_set "bulk" host APIs.
+ *
+ * The bulk APIs are only invocable from the host and are used for doing operations like `insert` or
+ * `contains` on a set of keys.
+ *
+ */
+int main(void)
+{
+  using Key = int;
+
+  // Empty slots are represented by reserved "sentinel" values. These values should be selected such
+  // that they never occur in your input data.
+  Key constexpr empty_key_sentinel = -1;
+
+  // Number of keys to be inserted
+  std::size_t constexpr num_keys = 50'000;
+
+  // Compute capacity based on a 50% load factor
+  auto constexpr load_factor = 0.5;
+  std::size_t const capacity = std::ceil(num_keys / load_factor);
+
+  // Constructs a set with at least `capacity` slots using -1 as the empty keys sentinel.
+  cuco::static_set<Key,
+                   cuco::extent<std::size_t>,
+                   cuda::thread_scope_device,
+                   thrust::equal_to<Key>,
+                   cuco::perfect_probing<4,  // CG size
+                                         cuco::default_hash_function<Key>>>
+    set{capacity, cuco::empty_key{empty_key_sentinel}};
```

Comment:
Sounds good!

Reply:
Remind me please.  Is this because creating the perfect prober is potentially impossible so we want to have a `create_perfect_probing` function that returns `std::optional<.....>`?

You could also make the constructor public and just have it throw in case of failure.  That's a normal way for constructors to fail.

Or is there another good reason?  I don't remember.